File: crates/dsp-runtime/src/protocol/audio.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct AudioEngineConfig {
    pub backend: String,
    pub input_device_id: String,
    pub output_device_id: String,
    pub buffer_size: u32,
    pub session_sample_rate: Option<u32>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct AudioBackend {
    pub id: String,
    pub label: String,
    pub available: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct AudioDevice {
    pub id: String,
    pub name: String,
    pub is_default: bool,
    pub default_sample_rate: Option<u32>,
    pub min_buffer_size: Option<u32>,
    pub max_buffer_size: Option<u32>,
    pub channel_count: Option<u32>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct AudioDeviceList {
    pub inputs: Vec<AudioDevice>,
    pub outputs: Vec<AudioDevice>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct AudioRuntime {
    pub state: String,
    pub requested_buffer_size: Option<u32>,
    pub sample_rate: Option<u32>,
    pub input_sample_rate: Option<u32>,
    pub output_sample_rate: Option<u32>,
    pub input_buffer_size: Option<u32>,
    pub output_buffer_size: Option<u32>,
    pub ring_buffer_capacity_frames: Option<u32>,
    pub ring_buffer_fill_frames: Option<u32>,
    pub input_latency_ms: Option<f64>,
    pub output_latency_ms: Option<f64>,
    pub ring_buffer_latency_ms: Option<f64>,
    pub engine_latency_ms: Option<f64>,
    pub estimated_round_trip_latency_ms: Option<f64>,
    pub xruns: u32,
    pub clock_sync: String,
    pub buffer_fallback: bool,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub struct RoundTripLatencyMeasurementRequest {
    pub input_channel: u32,
    pub output_channel: u32,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct RoundTripLatencyMeasurement {
    pub status: String,
    pub input_channel: Option<u32>,
    pub output_channel: Option<u32>,
    pub measured_round_trip_latency_ms: Option<f64>,
    pub failure: Option<String>,
}

File: packages/contracts/src/audio.ts
import type { ProjectLifecycleState } from "./project"
import type { RecordingLifecycleState } from "./recording"

// "mock" is a cpal custom host that synthesises capture and discards playback.
// It is always available and is listed last so it is only auto-selected when no
// hardware backend can be reached.
export const AUDIO_BACKENDS = ["wasapi", "asio", "coreaudio", "alsa", "mock"] as const
export type AudioBackend = (typeof AUDIO_BACKENDS)[number]

export interface AudioBackendDescriptor {
  id: AudioBackend
  label: string
  available: boolean
}

export interface AudioDeviceDescriptor {
  id: string
  name: string
  isDefault: boolean
  defaultSampleRate: number | null
  minBufferSize: number | null
  maxBufferSize: number | null
  channelCount: number | null
}

export interface AudioDeviceList {
  inputs: AudioDeviceDescriptor[]
  outputs: AudioDeviceDescriptor[]
}

export const AUDIO_BUFFER_SIZES = [32, 64, 128, 256, 512, 1024, 2048] as const
export type AudioBufferSize = number

export interface AudioPreferences {
  backend: AudioBackend
  inputDeviceId: string
  outputDeviceId: string
  bufferSize: AudioBufferSize
}

export const DEFAULT_AUDIO_PREFERENCES: AudioPreferences = {
  backend: "wasapi",
  inputDeviceId: "",
  outputDeviceId: "",
  bufferSize: 256
}

export type AudioEngineState = "stopped" | "running" | "error"
export type AudioClockSync = "inactive" | "shared-device" | "adaptive-resampled"

export interface AudioRuntimeSnapshot {
  state: AudioEngineState
  requestedBufferSize: number | null
  sampleRate: number | null
  inputSampleRate: number | null
  outputSampleRate: number | null
  inputBufferSize: number | null
  outputBufferSize: number | null
  ringBufferCapacityFrames: number | null
  ringBufferFillFrames: number | null
  inputLatencyMs: number | null
  outputLatencyMs: number | null
  ringBufferLatencyMs: number | null
  engineLatencyMs: number | null
  estimatedRoundTripLatencyMs: number | null
  xruns: number
  clockSync: AudioClockSync
  bufferFallback: boolean
}

export interface RoundTripLatencyMeasurementRequest {
  inputChannel: number
  outputChannel: number
}

export type RoundTripLatencyMeasurementStatus =
  "idle" | "preparing" | "measuring" | "complete" | "failed"

export type RoundTripLatencyMeasurementFailure = "input-too-loud" | "signal-not-detected"

export interface RoundTripLatencyMeasurement {
  status: RoundTripLatencyMeasurementStatus
  inputChannel: number | null
  outputChannel: number | null
  measuredRoundTripLatencyMs: number | null
  failure: RoundTripLatencyMeasurementFailure | null
}

export const INITIAL_AUDIO_RUNTIME_SNAPSHOT: Readonly<AudioRuntimeSnapshot> = {
  state: "stopped",
  requestedBufferSize: null,
  sampleRate: null,
  inputSampleRate: null,
  outputSampleRate: null,
  inputBufferSize: null,
  outputBufferSize: null,
  ringBufferCapacityFrames: null,
  ringBufferFillFrames: null,
  inputLatencyMs: null,
  outputLatencyMs: null,
  ringBufferLatencyMs: null,
  engineLatencyMs: null,
  estimatedRoundTripLatencyMs: null,
  xruns: 0,
  clockSync: "inactive",
  bufferFallback: false
}

export type AudioLifecycleState =
  | { status: "stopped"; runtime: AudioRuntimeSnapshot; error: string | null }
  | { status: "starting"; runtime: AudioRuntimeSnapshot; error: null }
  | { status: "running"; runtime: AudioRuntimeSnapshot; error: string | null }
  | { status: "reconfiguring"; runtime: AudioRuntimeSnapshot; error: null }
  | { status: "stopping"; runtime: AudioRuntimeSnapshot; error: null }
  | { status: "error"; runtime: AudioRuntimeSnapshot; error: string }

export interface DesktopLifecycleSnapshot {
  revision: number
  project: ProjectLifecycleState
  audio: AudioLifecycleState
  recording: RecordingLifecycleState
}

export type DesktopLifecycleEvent =
  | {
      type: "project"
      revision: number
      state: ProjectLifecycleState
    }
  | {
      type: "audio"
      revision: number
      state: AudioLifecycleState
    }
  | {
      type: "recording"
      revision: number
      state: RecordingLifecycleState
    }
